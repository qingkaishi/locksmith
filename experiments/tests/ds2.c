/* Data structure example 2: A non-recursive pair data structure.  The
   data in the pair is protected by the lock in the pair. */

#include <pthread.h>

struct pair {
  pthread_mutex_t lock;
  int data;
};

struct pair shared;

void *run(void *arg) {
  pthread_mutex_lock(&shared.lock);
  shared.data++;
  pthread_mutex_unlock(&shared.lock);
  return NULL;
}
int main() {
  pthread_t t1, t2;

  pthread_mutex_init(&shared.lock, NULL);
  pthread_create(&t1, NULL, run, NULL);
  pthread_create(&t2, NULL, run, NULL);
  run(0);
  return 1;
}
