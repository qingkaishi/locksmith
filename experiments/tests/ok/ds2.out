************************* STARTING *************************

shared: alloc.data:ds2.c:11
alloc.data:ds2.c:11 is protected by:
  concrete shared.lock:ds2.c:22
  shared.lock:ds2.c:11
*************************** DONE ***************************
